### DISCUSSION TITLE
ApplicationLifetime.Shutdown not fully shutting down process

### DISCUSSION THREAD

**Discussion by logiclrd** [AvaloniaUI/Avalonia#15904] [Q&A] [answered]
I have an application that I need to start up without any window initially visible. I am setting `ApplicationLifetime.ShutdownMode` to `OnExplicitShutdown` in the `OnFrameworkInitializationCompleted` handler of my `Application` subclass. Then, inside a `Command` execute handler attached to a `TrayIcon` menu item, I call `ApplicationLifetime.Shutdown` (casting through `IClassicDesktopStyleApplicationLifetime`). If I have a window open, the window disappears, but the tray icon remains and the process does not immediately exit. Some time later, the process exits with a weird exception such as:
```
System.InvalidOperationException: Call from invalid thread
   at Avalonia.Threading.Dispatcher.<VerifyAccess>g__ThrowVerifyAccess|16_0()
   at Avalonia.AvaloniaObject.GetValue[T](StyledProperty`1 property)
   at Avalonia.Controls.NativeMenuItem.get_Command()
   at Avalonia.Controls.NativeMenuItem.CanExecuteChanged()
   at Avalonia.Controls.NativeMenuItem.CanExecuteChangedSubscriber.OnEvent(Object sender, WeakEvent ev, EventArgs e)
   at Avalonia.Utilities.WeakEvent`2.Subscription.OnEvent(Object sender, TEventArgs eventArgs)
   at Avalonia.Utilities.WeakEvent.<>c__DisplayClass2_1`1.<Register>b__1(Object _, EventArgs e)
   at ReactiveUI.ReactiveCommandBase`2.OnCanExecuteChanged(Boolean newValue) in /_/src/ReactiveUI/ReactiveCommand/ReactiveCommandBase.cs:line 219
   at System.Reactive.AnonymousSafeObserver`1.OnNext(T value) in /_/Rx.NET/Source/src/System.Reactive/AnonymousSafeObserver.cs:line 43
   at System.Reactive.Sink`1.ForwardOnNext(TTarget value) in /_/Rx.NET/Source/src/System.Reactive/Internal/Sink.cs:line 49
...
(25 stack frames removed)
...
   at System.Reactive.Sink`1.ForwardOnNext(TTarget value) in /_/Rx.NET/Source/src/System.Reactive/Internal/Sink.cs:line 49
   at System.Reactive.Linq.ObservableImpl.Scan`2._.OnNext(TSource value) in /_/Rx.NET/Source/src/System.Reactive/Linq/Observable/Scan.cs:line 48
   at System.Reactive.SafeObserver`1.WrappingSafeObserver.OnNext(TSource value) in /_/Rx.NET/Source/src/System.Reactive/Internal/SafeObserver.cs:line 30
   at System.Reactive.Sink`1.ForwardOnNext(TTarget value) in /_/Rx.NET/Source/src/System.Reactive/Internal/Sink.cs:line 49
   at System.Reactive.ObserveOnObserverLongRunning`1.Drain() in /_/Rx.NET/Source/src/System.Reactive/Internal/ScheduledObserver.cs:line 732
   at System.Reactive.ObserveOnObserverLongRunning`1.<>c.<.cctor>b__17_0(ObserveOnObserverLongRunning`1 self, ICancelable cancelable) in /_/Rx.NET/Source/src/System.Reactive/Internal/ScheduledObserver.cs:line 680
   at System.Reactive.Concurrency.DefaultScheduler.LongRunning.LongScheduledWorkItem`1.<>c.<.ctor>b__3_0(Object thisObject) in /_/Rx.NET/Source/src/System.Reactive/Concurrency/DefaultScheduler.cs:line 181
   at System.Reactive.Concurrency.ConcurrencyAbstractionLayerImpl.<>c.<StartThread>b__8_0(Object itemObject) in /_/Rx.NET/Source/src/System.Reactive/Concurrency/ConcurrencyAbstractionLayerImpl.cs:line 75
```
Something is keeping the process alive, but I don't know how to find out what. Shouldn't `.Shutdown()` kill everything off and exit the process?

**Comment by maxkatz6** [comment]
> Shouldn't .Shutdown() kill everything off and exit the process?

No. `Shutdown` should gracefully close all the windows and stop the event loop. Closing application after.
If you want to `kill` the app, use `Environment.Exit(0)` for example.

> Something is keeping the process alive

Yes, something. Might be a bug either in the app or the framework.

**Reply by logiclrd** [reply]
> Yes, something. Might be a bug either in the app or the framework.

Do you have any ideas on how to debug this?

**Reply by DJGosnell** [reply]
I'd look into the threads running after you have tried to shutdown and pause it.  It'll give you more insight into what's going on on that thread that is holding the application open after the Shutdown.

**Reply by logiclrd** [reply]
I don't see any way to break execution other than a breakpoint. Maybe this is a Linux limitation? I see screenshots with a Pause button, but it doesn't show up on my screen.

I just ran a debugging session and stepped over `desktop.Shutdown()`, and after it returned, the only thread with any stack frames in my code was the one that just called `desktop.Shutdown()`. Every other frame just says "[External Code] Unknown Source".

Actually I just tried a second time and this time I got an `InvalidOperationException: Call from invalid thread`. The stack trace goes through a handful of Avalonia-specific frames and a _whole bunch_ of `System.Reactive.dll` frames, and the listing terminates with a `[Native to Managed Transition]`.

Based on this, it's probably not my code doing this?

**Comment by damian-666** [comment]
See below answer before doing this..   I get this way too often with **Microsoft developer studio** and i suspect its the **debugger holding the process open.  or something.**   then you have to **wait a while for it to appear in task manager, then hunt it as it moves, click and kill the process**.   that s after detaching the debugger so much be another  debugging server like an inspector.   **so i might suggest launching a little exe that kills Avalonia after a certain time so if it fails to shut down for whatever reason**, you can rebuild and debug and ignore the problem is might not be worth tracking and  solving, or in developer studio, or remember to press exit if you say stop debugging, or if debugger is detached can be detected, post yourself a ProcessExit call .   **but be careful not to leave unsaved data, leaked temp file and such.**      **if you are saving the data, and getting past closing windows asking if stuff needs be saved.. killing the process should not cause leaks, maybe temp files , which again can be cleaned up.  its ugly but  using so many complex sync technology on different platforms, well that's how you get it shipped..    i have so many leaked processes   in windows..**    its not that bad but when you are trying to build debug its horror. costs minutes every time since i don't even see my Avalonia or process in the task manager for a minute or two..  i get the leak like this in the debugger after exceptions uncaught maybe or if  **say stop debugging , without closing the application.**  I would dare say its a visual studio issue or **lack of feature, or it should offer to kill the process if it cannot get write access to it.**

**Comment by timunie** [comment]
Can this be reproduced in a minimal sample?

**Reply by logiclrd** [reply]
I will see what I can do. :-)

**Reply by logiclrd** [reply, marked as answer]
Welp, I can no longer reproduce the problem. I have a thread that pumps notification messages, and it exits when the application is shutting down (confirmed with debug output). But, I figured, what the hey, let's be semantically correct about things, and marked it `IsBackground = true`, and now when I call `ApplicationLifetime.Shutdown`, after getting the aforementioned exception from something internal to Avalonia ("Call from invalid thread"), the process terminates. I'm pretty sure that's the only semantic change I have made since I last observed the problem. So apparently, even though my thread's method had returned and there was nothing in any running thread's stack trace owned by me, it was still somehow sticking around and keeping the process alive. Weird.

**Reply by timunie** [reply]
Could also be the debugger. So make sure to test ot without debugger. Also if you manage to reproduce, attach the minimum sample here. 

Last but not least worth to check 11.1 beta for possible fixes already in

**Reply by logiclrd** [reply]
Awesome, thanks so much :-) Will come back if it starts happening again.

**Reply by damian-666** [reply]
if you are using **visual studio i**  this is probably the problem , fix workaround is handling app top level exceptions and/ or Shift F5, stop debugging , and  not uncommon for problems to remain unaddressed for 11 years. 
**https://stackoverflow.com/questions/11140525/visual-studio-killing-specific-processes-on-stop-debug-shiftf5**

  i always "stop debugging".   I don't want to press app close first, so  i like the custom workaround by adding that script, and put that right on your visual studio menu next to Stop debugging ( remap shift f5 to that custom command) its hard  to break this habit. 
   
as  next person said it can happen when you ship,  and so  then you can catch all exceptions at the app level, and then  and flush all files, output it carefully as the UI thread could be tied up, make a temp recovery file of any unsaved user data,  or eat it . and provide a warning on next startup if they want to recover the file .    the other hacky workaround, see  if there is no kill current process feature or command hidden  in dev studio.  But the hack solution , is still needed on a dirty exit, or  if  it hung in thread lock , or in an  infinite loop  .    it doesn't always happen, but  often.  takes too long to recover via task manager.
